flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergSink.java

```diff
@@ -0,0 +1,761 @@
+
+    public Builder snapshotProperties(Map<String, String> properties) {
+      snapshotSummary.putAll(properties);
+      return this;
+    }
+
+    public Builder setSnapshotProperty(String property, String value) {
+      snapshotSummary.put(property, value);
+      return this;
+    }
+
+    public Builder toBranch(String branch) {
+      writeOptions.put(FlinkWriteOptions.BRANCH.key(), branch);
+      return this;
+    }
+
+    private String operatorName(String suffix) {
+      return uidPrefix != null ? uidPrefix + "-" + suffix : suffix;
+    }
+
+    public IcebergSink build() {
+
+      Preconditions.checkArgument(
+          inputCreator != null,
+          "Please use forRowData() or forMapperOutputType() to initialize the input DataStream.");
+      Preconditions.checkNotNull(tableLoader(), "Table loader shouldn't be null");
+
+      // Set the table if it is not yet set in the builder, so we can do the equalityId checks
+      this.table = checkAndGetTable(tableLoader(), table);
+
+      // Validate the equality fields and partition fields if we enable the upsert mode.
+      List<Integer> equalityFieldIds =
+          SinkUtil.checkAndGetEqualityFieldIds(table, equalityFieldColumns);
+
+      // Init the `flinkWriteConf` here, so we can do the checks
+      FlinkWriteConf flinkWriteConf = new FlinkWriteConf(table, writeOptions, readableConfig);
+      String branch = flinkWriteConf.branch();
+      boolean overwriteMode = flinkWriteConf.overwriteMode();
+
+      if (flinkWriteConf.upsertMode()) {
+        Preconditions.checkState(
+            !flinkWriteConf.overwriteMode(),
+            "OVERWRITE mode shouldn't be enable when configuring to use UPSERT data stream.");
+        Preconditions.checkState(
+            !equalityFieldIds.isEmpty(),
+            "Equality field columns shouldn't be empty when configuring to use UPSERT data stream.");
+        if (!table.spec().isUnpartitioned()) {
+          for (PartitionField partitionField : table.spec().fields()) {
+            Preconditions.checkState(
+                equalityFieldIds.contains(partitionField.sourceId()),
+                "In UPSERT mode, partition field '%s' should be included in equality fields: '%s'",
+                partitionField,
+                equalityFieldColumns);
+          }
+        }
+      }
+
+      Map<String, String> writeProperties =
+          writeProperties(table, flinkWriteConf.dataFileFormat(), flinkWriteConf);
+      RowType flinkRowType = toFlinkRowType(table.schema(), tableSchema);
+
+      Duration tableRefreshInterval = flinkWriteConf.tableRefreshInterval();
+      SerializableSupplier<Table> tableSerializableSupplier;
+
+      if (tableRefreshInterval != null) {
+        tableSerializableSupplier =
+            new CachingTableSupplier(
+                (SerializableTable) table, tableLoader(), tableRefreshInterval);
+      } else {
+        SerializableTable serializableTable = (SerializableTable) SerializableTable.copyOf(table);
```

Comment:
`table` type is `Table`, not `SerializableTable`. 
I've made `checkAndGetTable` to return `SerializableTable` but we still need the type casting when we instantiate `tableSupplier`
We could do more refactor in the code to avoid the type cast, but IMO this would introduce additional complexity which is not needed

Reply:
why can't `table` type be `SerializableTable`? what's the problem with instantiating `tableSupplier`?